numpy/core/arrayprint.py

```diff
@@ -610,30 +605,38 @@ def _formatArray(a, format_function, rank, max_line_len,
 
 class FloatFormat(object):
     def __init__(self, data, precision, suppress_small, sign=False):
+        # for backcompatibility, accept bools
+        if isinstance(sign, bool):
+            sign = '+' if sign else '-'
+
+        self._legacy = False
+        if sign == 'legacy':
+            self._legacy = True
+            sign = '-' if data.shape == () else ' '
+
         self.precision = precision
         self.suppress_small = suppress_small
         self.sign = sign
         self.exp_format = False
         self.large_exponent = False
-        self.max_str_len = 0
-        try:
-            self.fillFormat(data)
-        except (TypeError, NotImplementedError):
-            # if reduce(data) fails, this instance will not be called, just
-            # instantiated in formatdict.
-            pass
+
+        self.fillFormat(data)
 
     def fillFormat(self, data):
         with errstate(all='ignore'):
-            special = isnan(data) | isinf(data)
+            hasinf = isinf(data)
+            special = isnan(data) | hasinf
             valid = not_equal(data, 0) & ~special
-            non_zero = absolute(data.compress(valid))
+            non_zero = data[valid]
+            abs_non_zero = absolute(non_zero)
             if len(non_zero) == 0:
                 max_val = 0.
                 min_val = 0.
+                min_val_sgn = 0.
             else:
-                max_val = maximum.reduce(non_zero)
-                min_val = minimum.reduce(non_zero)
+                max_val = np.max(abs_non_zero)
+                min_val = np.min(abs_non_zero)
+                min_val_sgn = np.min(non_zero)
```

Comment:
Seems a bit odd that we're importing `np` directly for these, yet importing other stuff indirectly. Not something you need to fix in this PR though

Reply:
Yes, looks like we can replace all the indirect imports can be replaced by the single `np` import - I just tried it, seems fine.

I won't do that in this PR though.

Reply:
I've never understood quite how, or why, that worked. But it generally does.

Reply:
I think it works because we never use or call any numpy functions or variables in this file at import time, only at runtime. 

At the moment `arrayprint` is imported from the main`__init__` file, the `numpy` module has been created but not populated, so we are allowed to import numpy, but not allowed refer to any of its functions or variables. However, we can use it inside of function definitions since by the time they are called the numpy module will have been populated.

(Actually, there I see exceptions: At module import time we call `multiarray.set_string_function` and we refer to `int_`, `float_`, `complex_`, `intc` and `longlong`. Therefore these will still need to be directly imported).

Reply:
Hmm, but it might be safer to leave these all the way they are: What if another module calls arrayprint functions at import time? Then we would have a problem. Ideally no other modules would, though.

(And this isn't theoretical: We recently had a problem because `getlimits.py`was calling `array2string` at module import time, though we've now fixed it so it no longer does (#9113))